module-apps/application-call/windows/CallWindow.cpp

```diff
@@ -299,34 +299,61 @@ namespace gui
 
     void CallWindow::handleAudioEvent(const audio::Event &event)
     {
-
-        switch (event.getType()) {
-        case audio::EventType::BlutoothHFPDeviceState:
-        case audio::EventType::BlutoothHSPDeviceState:
+        const auto eventType = event.getType();
+        switch (eventType) {
+        case audio::EventType::BluetoothHFPDeviceState:
+        case audio::EventType::BluetoothHSPDeviceState:
         case audio::EventType::JackState: {
             if (event.getDeviceState() == audio::Event::DeviceState::Connected) {
-                devices_connected.insert(event.getType());
-                changeSpeakerIconIfNeeded();
+                handleConnectAudioEvent(eventType);
             }
             else {
-                devices_connected.erase(event.getType());
-                if (not devices_connected.empty()) {
-                    changeSpeakerIconIfNeeded();
-                }
+                handleDisconnectAudioEvent(eventType);
             }
             break;
         }
+        case audio::EventType::CallLoudspeakerOn: {
+            isLouderSpeakerON = true;
+            break;
+        }
+        case audio::EventType::CallLoudspeakerOff: {
+            isLouderSpeakerON = false;
+            break;
+        }
         default: {
             break;
         }
         }
     }
 
-    void CallWindow::changeSpeakerIconIfNeeded()
+    void CallWindow::handleConnectAudioEvent(const audio::EventType &eventType)
+    {
+        devices_connected.insert(eventType);
+        turnOffSpeakerIconIfNeeded();
+    }
+
+    void CallWindow::handleDisconnectAudioEvent(const audio::EventType &eventType)
+    {
+        devices_connected.erase(eventType);
+        if (not devices_connected.empty()) {
+            turnOffSpeakerIconIfNeeded();
+        }
+        else {
+            turnOnSpeakerIconIfNeeded();
+        }
+    }
+
+    void CallWindow::turnOffSpeakerIconIfNeeded()
```

Comment:
Same as above, ought to be implemented inside presenter, not in the window class. By doing this, we are following fundamental rule of SRP(Single responsibility rule), i.e. window(view) is responsible only for displaying data, presenter is responsible for business logic and orchestration between models and view and so on.